Charts: Add interactive legend support to Pie and Semi-Circle charts

annacmc opened the pull request:
## Proposed changes:
* Add `legendInteractive` prop to PieChart component to enable clickable legend items
* Add `legendInteractive` prop to PieSemiCircleChart component 
* Implement segment visibility toggling when legend items are clicked
* Add percentage recalculation logic so visible segments always total 100%
* Maintain consistent segment colors when toggling visibility using original indices
* Display empty state message when all segments are hidden
* Update type definitions and JSDoc documentation for pie chart support
* Pass `interactive` prop to Legend component when `legendInteractive` is enabled
* Create smart legend data mapping that shows recalculated percentages for visible segments only
* Keep all legend items visible (styled as inactive) rather than removing hidden items

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?

No, it does not.

## Testing instructions:

### PieChart:
* Go to Storybook -> Charts -> Types -> Pie Chart -> Interactive Legend story
* Click on legend items to toggle segment visibility
* Verify segments show/hide smoothly
* Check that visible segment percentages recalculate to total 100%
* Example: Hide a 25% segment, remaining segments should show 66.67% and 33.33%
* Verify hidden legend items remain visible but styled as inactive (grayed out)
* Hide all segments and verify empty state message appears
* Click hidden items again to show segments
* Verify segment colors remain consistent when toggling

### DonutChart:
* Go to Storybook -> Charts -> Types -> Pie Chart -> Donut Interactive Legend story
* Test same functionality as above with donut variant
* Verify the donut hole remains visible in empty state

### PieSemiCircleChart:
* Go to Storybook -> Charts -> Types -> Pie Semi Circle Chart -> Interactive Legend story
* Test same interactive legend functionality
* Verify semi-circle layout is maintained when segments are hidden
* Check percentage recalculation works correctly

### All Charts:
* Verify clicking legend items requires `legendInteractive` prop to be true
* Verify `chartId` and `GlobalChartsProvider` are required for functionality
* Check that aria-pressed attribute updates correctly (true/false)
* Test keyboard accessibility (Enter and Space keys should toggle)
* Verify no TypeScript errors
* Confirm no visual regressions in tooltips, labels, or other features
* Run tests: `npm test -- src/components/pie-chart/test`
* Run tests: `npm test -- src/components/pie-semi-circle-chart/test`

annacmc committed:
Update legendInteractive JSDoc to include pie chart support

annacmc committed:
Add interactive legend support to PieChart

annacmc committed:
Add tests for PieChart interactive legend functionality

annacmc committed:
Add interactive legend story for PieChart

annacmc committed:
Add interactive legend story for DonutChart

annacmc committed:
Document legendInteractive prop in PieChart API reference

annacmc committed:
Add interactive legend support to PieSemiCircleChart

annacmc committed:
Add tests for PieSemiCircleChart interactive legend

annacmc committed:
Add interactive legend story for PieSemiCircleChart

annacmc committed:
Document legendInteractive prop in PieSemiCircleChart API reference

annacmc committed:
changelog

annacmc committed:
Fix padAngle calculation to use original data length for consistent spacing

kangzj reviewed (commented):
1. We should hide the text in the center too when there's no data:

<img width="531" height="550" alt="image" src="https://github.com/user-attachments/assets/953a12a4-c4ae-492c-8e7a-98356df1c995" />


2. Setting whether the legend is interactive shouldn't change the format of legend lable:

https://github.com/user-attachments/assets/2ffac75b-24c2-45e9-a611-23db16917805

3. LegendInteractive doesn't work for composition legend stories (we might just remove the option):

http://localhost:50240/?path=/story/js-packages-charts-types-pie-chart--with-composition-legend

kangzj commented on projects/js-packages/charts/src/components/pie-semi-circle-chart/pie-semi-circle-chart.tsx:
```diff
@@ -183,7 +192,35 @@ const PieSemiCircleChartInternal: FC< PieSemiCircleChartProps > = ( {
+	const { getElementStyles, isSeriesVisible } = useGlobalChartsContext();
+
+	// Filter and recalculate data for interactive legends
+	const visibleData = useMemo( () => {
```
Feels we could consolidate with https://github.com/Automattic/jetpack/pull/45580/files#diff-0d8c97636a7c1ee076e829b87c6e323bc97e6f646592f949b9b88d6bf0df89b0R186

annacmc committed:
Hide center text when all segments hidden and document interactive legend limitation

annacmc committed:
Hide legendInteractive control in composition API stories

annacmc committed:
Extract interactive legend data logic into shared useInteractiveLegendData hook

Create a reusable custom hook that consolidates duplicate logic between
PieChart and PieSemiCircleChart components. The hook handles:
- Filtering data based on legend visibility
- Recalculating percentages for visible segments
- Tracking empty state when all segments are hidden

This follows the DRY principle and makes the codebase more maintainable.

kangzj commented on projects/js-packages/charts/src/components/pie-chart/pie-chart.tsx:
```diff
@@ -171,14 +180,45 @@ const PieChartInternal = ( {
-	const legendItems = useChartLegendItems( data, legendOptions );
+	// Always use original data for legend - the Legend component handles visibility styling
+	// For interactive legends, we need to show recalculated percentages for visible items only
+	const legendData = useMemo( () => {
```
This should be unified too

https://github.com/Automattic/jetpack/pull/45580/files#diff-787f983bad6d02257d680889ba44311cbd927aabe21003f40fadc45b3933c723R228

kangzj reviewed (dismissed):
feel free to ship after addressing this: https://github.com/Automattic/jetpack/pull/45580/files#r2454038480

annacmc committed:
Add legendData to useInteractiveLegendData hook for complete DRY consolidation

Extend the hook to also handle legend data preparation with recalculated
percentages. This eliminates another 36 lines of duplicate code between
PieChart and PieSemiCircleChart components.

The hook now returns:
- visibleData: Filtered chart data for rendering
- allSegmentsHidden: Empty state flag
- legendData: All data with recalculated percentages for visible items

This completes the consolidation of all interactive legend logic.
